lib/auth/password.go

```diff
@@ -10,13 +10,61 @@ import (
 	"github.com/gravitational/teleport/lib/services"
 	"github.com/gravitational/teleport/lib/utils"
 	"github.com/gravitational/trace"
+	"github.com/tstranex/u2f"
 
 	"github.com/pquerna/otp"
 	"github.com/pquerna/otp/totp"
 )
 
 var fakePasswordHash = []byte(`$2a$10$Yy.e6BmS2SrGbBDsyDLVkOANZmvjjMR890nUGSXFJHBXWzxe7T44m`)
 
+// ChangePasswordWithTokenRequest defines a request to change user password
+type ChangePasswordWithTokenRequest struct {
+	// SecondFactorToken is 2nd factor token value
+	SecondFactorToken string `json:"second_factor_token"`
+	// TokenID is this token ID
+	TokenID string `json:"token"`
+	// Password is user password
+	Password []byte `json:"password"`
+	// U2FRegisterResponse is U2F register response
+	U2FRegisterResponse u2f.RegisterResponse `json:"u2f_register_response"`
+}
+
+// ChangePasswordWithToken changes password with user token
+func (s *AuthServer) ChangePasswordWithToken(req ChangePasswordWithTokenRequest) (services.WebSession, error) {
+	user, err := s.changePasswordWithToken(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sess, err := s.createUserWebSession(user)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return sess, nil
+}
+
+// ResetPassword securely generates a new random password and assigns it to user
```

Comment:
The comment should explain that this function is used to invalidate a user during the password reset process.